### PR TITLE
Universal IntersectionObserver polyfill for cross-origin iframes

### PR DESCRIPTION
**Pull request by dvoytenko**
Partial for #27807.

### CODE REVIEW

File: extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js [hunk 1]
```diff
@@ -27,7 +28,9 @@ const TAG = 'amp-intersection-observer-polyfill';
 function upgrade(win) {
   upgradePolyfill(win, () => {
     installIntersectionObserver();
-    // TODO(dvoytenko): hookup 3p host updater for inabox.
+    if (win != win.top) {
```
Comment (zhouyx) [inline extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js:31]: I have a question about this. Not sure how `win.top` work, does this not cover FIE case. 
Also I know the AMP doc can be served from a full screen iframe. Will that affect the check here?

File: src/utils/intersection-observer-polyfill.js [hunk 2]
```diff
@@ -100,8 +101,15 @@ export function getIntersectionChangeEntry(element, owner, hostViewport) {
  * @return {boolean}
  */
 export function nativeIntersectionObserverSupported(win) {
+  if (!('IntersectionObserver' in win)) {
+    return false;
+  }
+  // eslint-disable-next-line no-undef
+  if (INTERSECTION_OBSERVER_POLYFILL || getMode().localDev || getMode().test) {
```
Comment (zhouyx) [inline src/utils/intersection-observer-polyfill.js:108]: nativeInOb is considered supported even when the connection between inabox and its host failed. 
I am a bit worried that it will change the behavior in that case. 
https://github.com/ampproject/amphtml/blob/97099ce8bf4826a08d964e7a02da55a7641b1090/extensions/amp-analytics/0.1/visibility-manager.js#L761-L767
the old polyfill tick with `this.vewport_getRect()` once, what's the behavior with INTERSECTION_OBSERVER_POLYFILL inabox host connection failure?

File: extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js (hunk 1 above)
Comment (zhouyx) [inline extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js:31]: I have a question about `win.top`. I know the AMP doc can be served from a full screen iframe. Will that affect the check here?

File: src/utils/intersection-observer-polyfill.js [hunk 3]
```diff
@@ -100,8 +101,15 @@ export function getIntersectionChangeEntry(element, owner, hostViewport) {
...
+    // For the new stub polyfill it's enough to have a stub to be functional.
+    return true;
+  }
```
Comment (jridgewell) [inline src/utils/intersection-observer-polyfill.js:111]: Why is this change needed?

File: extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js [hunk 4]
```diff
@@ -27,7 +28,9 @@ const TAG = 'amp-intersection-observer-polyfill';
...
+      maybeSetupCrossOriginObserver(win);
```
Comment (jridgewell) [inline extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js:32]: Won't this bring the cross-origin listener into every page? Eg, it'll now install an inabox listener for regular amp pages.

File: src/utils/intersection-observer-polyfill.js (hunk 2 above)
Comment (dvoytenko) [reply src/utils/intersection-observer-polyfill.js:108]: The default behavior: all observers are blocked. It's very easy to unblock them, e.g. when connection fails. It should just go the polyfill extension. We just need to figure out the timeout. Do we really want 0 timeout for connection issues?

File: extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js (hunk 1 above)
Comment (dvoytenko) [reply extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js:31]: No, should not have any effect on fullscreen mode. This check overall not even necessary. But I felt like it reads cleaner: "do this only when iframed".

File: src/utils/intersection-observer-polyfill.js (hunk 3 above)
Comment (dvoytenko) [reply src/utils/intersection-observer-polyfill.js:111]: Technically, the polyfill doesn't define the prototype for `IntersectionObserverEntry`. E.g. this will fail: ` 'intersectionRatio' in win.IntersectionObserverEntry.prototype`. That being said, post-launch, this whole method and all downstream code will be gone, so I figured I don't need to press this issue.

File: extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js (hunk 4 above)
Comment (dvoytenko) [reply extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js:32]: No. `maybeSetupCrossOriginObserver` takes care of this by checking we are actually running in a host-enabled AMP iframe such as inabox. The details of how that happens are kind of messy, so I wanted to keep it hidden inside the `maybeSetupCrossOriginObserver` method. This also makes it easier to test.

File: extensions/amp-intersection-observer-polyfill/0.1/cross-origin-observer.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import {layoutRectLtwh} from '../../../src/layout-rect';
+
+/**
+ * Setup cross-origin iframe polyfill for AMP iframes, such as inabox.
+ *
+ * @param {!Window} win
+ */
+export function maybeSetupCrossOriginObserver(win) {
+  // Cross-origin iframe polyfill for AMP iframes, such as inabox.
+  const iframeClient = iframeMessagingClientFor(win);
```
Comment (zhouyx) [inline extensions/amp-intersection-observer-polyfill/0.1/cross-origin-observer.js:28]: `iframeClient` is the check to setup cross origin observer. Should we rely on `getMode.(win).runtime` instead to be more specific?

File: extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js (hunk 1 above)
Comment (zhouyx) [reply extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js:31]: I see. a bit confusing though, saw Justin had the question : P

File: build-system/test-configs/dep-check-config.js [hunk 6]
```diff
@@ -169,6 +169,9 @@ exports.rules = [
   {
     filesMatching: '{src,extensions}/**/*.js',
     mustNotDependOn: '3p/**/*.js',
+    allowlist: [
+      'src/inabox/inabox-iframe-messaging-client.js->3p/iframe-messaging-client.js',
```
Comment (zhouyx) [inline build-system/test-configs/dep-check-config.js:173]: curious about the size of the intersection-observer-polyfill extension.

File: src/utils/intersection-observer-polyfill.js (hunk 2 above)
Comment (zhouyx) [reply src/utils/intersection-observer-polyfill.js:108]: Checked the code again. LGTM, the reason we want tick with 0 timeout is also to polyfill the `observe` method, to invoke the callback when it is registered. 
It should change the behavior here because the initial `inaboxViewport.boxRect_` is placed below the window, so that the visibility event won't be triggered.

This is the ideal solution based on the comment here
https://github.com/ampproject/amphtml/blob/d94fa740ed6f9ab098248b207f5c161654752ebe/src/inabox/inabox-viewport.js#L457-L460

Comment (dvoytenko) [reply src/utils/intersection-observer-polyfill.js:108]: The InOb polyfill is definitely more desirable. But could you clarify? Should we timeout host connection here? I.e. would it be better to report wrong values or not report them at all? And if it's a timeout thing, what should it be?

File: extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js (hunk 1 above)
Comment (dvoytenko) [reply extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js:31]: Yep. 2 out of 3 people got confused - it's definitely a signal to rework the code. I will do so.

File: extensions/amp-intersection-observer-polyfill/0.1/cross-origin-observer.js (hunk 5 above)
Comment (dvoytenko) [reply extensions/amp-intersection-observer-polyfill/0.1/cross-origin-observer.js:28]: Sure. I can add it as well. Though I thought technically, if `iframeClient` is available - it doesn't matter what type of runtime we are in.

File: build-system/test-configs/dep-check-config.js (hunk 6 above)
Comment (dvoytenko) [reply build-system/test-configs/dep-check-config.js:173]: This particular change is actually completely unrelated to my change :) We are already using 3p/ in the src/inabox/. Somehow my change just triggered this dep check.

But the InOb polyfill itself is ~4K. We maybe adding an extra 0.5K or so to plug it all in.

File: extensions/amp-intersection-observer-polyfill/0.1/cross-origin-observer.js (hunk 5 above)
Comment (zhouyx) [reply extensions/amp-intersection-observer-polyfill/0.1/cross-origin-observer.js:28]: Just realized this is not always the case. inabox can be served in friendly iframe as well. 
https://github.com/ampproject/amphtml/blob/c619be8c5751243be74df8847cfb176b6e20872d/src/inabox/inabox-viewport.js#L481-L483

I believe we need to check against `isFriendlyIframed_` as well.

File: src/utils/intersection-observer-polyfill.js (hunk 2 above)
Comment (zhouyx) [reply src/utils/intersection-observer-polyfill.js:108]: give me one sec to verify the behavior when `visiblePercentageMin=0`

Comment (zhouyx) [reply src/utils/intersection-observer-polyfill.js:108]: IMO we don't need to timeout unless we want to do error reporting. 

I do think it's better to not report compared to report wrong values. That's the case today already with the only caveat that an event `[visiblePercentageMin: 0, visiblePercentageMax: 0]` (element is not visible) will still be triggered when the connection fails. And this PR will fix it.

File: extensions/amp-intersection-observer-polyfill/0.1/cross-origin-observer.js (hunk 5 above)
Comment (dvoytenko) [reply extensions/amp-intersection-observer-polyfill/0.1/cross-origin-observer.js:28]: This code is only be executing when loaded in a specific context - the iframe. And only when iframeClient is also set up. I think at least one of these should always be `false` for a friendly iframe case. No?

File: src/utils/intersection-observer-polyfill.js (hunk 2 above)
Comment (dvoytenko) [reply src/utils/intersection-observer-polyfill.js:108]: Ok. Sounds like this is resolved then.

File: extensions/amp-intersection-observer-polyfill/0.1/cross-origin-observer.js (hunk 5 above)
Comment (dvoytenko) [reply extensions/amp-intersection-observer-polyfill/0.1/cross-origin-observer.js:28]: For instance, at least [here](https://github.com/ampproject/amphtml/blob/90f1f7fcff01c68d703140e1cb6f6da2ebbabbef/src/inabox/inabox-iframe-messaging-client.js#L40) I see that iframe client is not setup for a friendly case. If my reading of this is right.

File: extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js (hunk 1 above)
Comment (dvoytenko) [reply extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js:31]: Reworked. I think it's a bit better now. PTAL.

File: extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js (hunk 4 above)
Comment (dvoytenko) [reply extensions/amp-intersection-observer-polyfill/0.1/amp-intersection-observer-polyfill.js:32]: Reworked to put it all in the `maybeSetupCrossOriginObserver` and added more docs there. I that's a little clearer now.

File: extensions/amp-intersection-observer-polyfill/0.1/cross-origin-observer.js (hunk 5 above)
Comment (zhouyx) [reply extensions/amp-intersection-observer-polyfill/0.1/cross-origin-observer.js:35]: Ohhh, I didn't know that the iframe client is only set up when in friendly iframe. You're absolute right about checking `iframeClient` instead of `getMode(win).runtime`! LGTM

File: src/utils/intersection-observer-polyfill.js (hunk 2 above)
Comment (zhouyx) [inline src/utils/intersection-observer-polyfill.js:108]: Since we won't be able to run experiment with inabox the same time as AMP docs, what about introducing a different experiment flag for the inabox related changes only. 
For example `INTERSECTION_OBSERVER_POLYFILL_INABOX`, then you can specify the default value of the const to `false` [here](https://github.com/ampproject/amphtml/blob/master/build-system/global-configs/experiments-const.json). 
Once we are about to run an experiment with inabox, you can then declare an experiment slot similar to what you have
https://github.com/ampproject/amphtml/blob/ab03e7c87443b5c1f69d4de1e409202475f5b238/build-system/global-configs/experiments-config.json#L2-L8

Comment (zhouyx) [reply src/utils/intersection-observer-polyfill.js:108]: caveat is that you won't get the e2e test coverage.

File: extensions/amp-intersection-observer-polyfill/0.1/cross-origin-observer.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+export function maybeSetupCrossOriginObserver(win) {
+  if (win == WindowInterface.getTop(win) || getMode(win).runtime != 'inabox') {
+    // Not an iframe at all.
+    return;
+  }
+
+  // Check if there is an iframe client connected and if it's indeed a polyfill.
+  const iframeClient = iframeMessagingClientFor(win);
+  const setupPolyfillUpdater =
+    win.IntersectionObserver['_setupCrossOriginUpdater'];
```
Comment (jridgewell) [inline extensions/amp-intersection-observer-polyfill/0.1/cross-origin-observer.js:37]: Where is `_setupCrossOriginUpdater` at?

Comment (dvoytenko) [reply extensions/amp-intersection-observer-polyfill/0.1/cross-origin-observer.js:37]: It's defined in the polyfill [here](https://github.com/w3c/IntersectionObserver/blob/60abaeacca6ceb039161863f24c98bcde8aeebfc/polyfill/intersection-observer.js#L172).

Comment (dvoytenko) [reply extensions/amp-intersection-observer-polyfill/0.1/cross-origin-observer.js:37]: And I can confirm: the declaration and caller are both compiled correctly.

File: src/utils/intersection-observer-polyfill.js (hunk 2 above)
Comment (dvoytenko) [reply src/utils/intersection-observer-polyfill.js]: Created a new constant for inabox. PTAL.

File: src/friendly-iframe-embed.js [hunk 8]
```diff
@@ -852,8 +852,14 @@ function installPolyfillsInChildWindow(parentWin, childWin) {
   // The anonymous class parameter allows us to detect native classes vs
   // transpiled classes.
   installCustomElements(childWin, class {});
-  // eslint-disable-next-line no-undef
-  if (INTERSECTION_OBSERVER_POLYFILL || getMode().localDev || getMode().test) {
+  if (
+    // eslint-disable-next-line no-undef
+    INTERSECTION_OBSERVER_POLYFILL ||
+    // eslint-disable-next-line no-undef
+    INTERSECTION_OBSERVER_POLYFILL_INABOX ||
```
Comment (zhouyx) [inline src/friendly-iframe-embed.js:859]: Does `INTERSECTION_OBSERVER_POLYFILL_INABOX` affect the FIE case?

Comment (dvoytenko) [reply src/friendly-iframe-embed.js:859]: Good catch. Not really. These use cases do not overlap. So I'm going to take it out to avoid confusion.
